### ISSUE TITLE
Issue with custom hook

### ISSUE THREAD

**Issue by adjourn** [jhonnymichel/react-hookstore#6]
I've discovered a weird issue:

If I use `useState` in custom hook in order to have my own reusable logic to `setState` and if I use `useState` somewhere else in app, `useMemo` inside custom hook is triggered twice. Not only that, each use of `useState` in app adds an invocation. Im not sure if I do something wrong or it's a bug. Any idea?

#### [CodeSandbox](https://codesandbox.io/s/wnz1zmnlok)

**Comment by adjourn** [comment]
Found the issue. I assumed that `set*` functions can be used as `set*(previousState => newState)` like React's `useState` but it's very buggy. Im not sure if it's supported use case (didn't find anything in documentation) but there you go.

Not only custom hooks are affected, every `set*` function that uses function updater seems to have the same bug, my theory: something cause several invocations, first invocation updates the store which then updates active views but other invocation(s) manage to silently update store as well and every following view seems to use that "wrong and silent" state.

**Comment by jhonnymichel** [comment]
Hi @adjourn, thanks for opening this issue! This is a very interesting problem.

Indeed, I have not supported passing a function to the store.setState method, this is the root of the problem. I overlooked this feature while trying to mimic React.useState API. My fault;

But since the library support reducers, I'm not sure I should add this or not. What do you think?

About the useMemo usage, since the store.setState method will dispatch the change to every component before you actually call useMemo, it is ineffective. I have plans on supporting memoization inside the library in upcoming updates.

since useMemo can not be called outside components, to make use of memoization for now, we need to use other implementations, such as our own or lodash's.

**To summarize**:
1. Your custom logic to update state should be moved to a reducer inside your store (not really required, but it would be the usual approach for this library)
2. Use Lodash or other custom memoization to avoid calling state update when not needed (I understand this should be built in on the library and I have plans on doing that).

Here is my fork of your code sample with memoization working (using Lodash.memoize) and custom logic moved to a store reducer: https://codesandbox.io/s/pw29rk6l9m

Please let me know what you think, best regards!

**Comment by adjourn** [comment]
I don't mind if function updater feature is not added, like you said, reducers are supported and state value can also be used in most common cases. I guess the biggest reason to implement it would be to have similar API with React - like I did assume that it's supported and Im probably not going to be the last one.

I agree that it would be convenient if memoization was built-in feature but I don't mind `Lodash` for now, it's already in my bundle anyway. Thanks for the library, really enjoying the API.

**Comment by adjourn** [comment]
I don't mind if function updater feature is not added, like you said, reducers are supported and state value can also be used in most common cases. I guess the biggest reason to implement it would be to have similar API with React - like I did assume that it's supported and Im probably not going to be the last one.

I agree that it would be convenient if memoization was built-in feature but I don't mind `Lodash.memoize` for now, it's already in my bundle anyway. Thanks for the library, really enjoying the API.
